### PR TITLE
Add Braille and Tones extension points to avoid monkey patching by API consumers

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Replaces #9917 

### Summary of the issue:
Tools like NVDA remote that need to intercept speech and braille output currently rely on monkeypatching to do this.

### Description of user facing changes
It is no longer possible to override the display size of the braille handler by setting braille.handler.displaySize. It is also no longer possible to enable/disable the handler by setting braille.handler.enabled.

### Description of development approach
Added the following extension points:

* inputCore.manager.decide_executeGesture: Decider for filtering gestures
* tones.decide_beep
* nvwave.decide_playWaveFile
* braille.handler.pre_writeCells
* braille.handler.filter_displaySize
* braille.handler.decide_enabled
* braille.handler.displayChange
* braille.handler.displaySizeChanged

This means that NVDA Remote and similar tools can at least drop all monkeypatching related to braille, playing tones and waves. Speech is a separate subject and will be handled in a follow up.

### Testing strategy:
* Tested that NVDA will still work as normal.
* Tested that lowering the display size programatically adds a full cell to the end of the text to indicate that there's a hard stop because of a lower cell count used.
* New unit tests

### Known issues with pull request:
None known

### Change log entries:
For Developers
- Added the following extension points: (Check them in source documentation for details).
 - inputCore.manager.decide_executeGesture
 - tones.decide_beep
 - nvwave.decide_playWaveFile
 - braille.handler.pre_writeCells
 - braille.handler.filter_displaySize
 - braille.handler.decide_enabled
 - braille.handler.displayChanged
 - braille.handler.displaySizeChanged

API breaking changes
- it is no longer possible to enable/disable the braille handler by setting braille.handler.enabled or update the display size of the handler by setting braille.handler.displaySize.
 - To update the displaySize programatically, register a handler to braille.handler.filter_displaySize. See the brailleViewer code for an example on how to do this.
 - To disable the braille handler programatically, register a handler to braille.handler.decide_enabled.

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [ ] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by CyrilleB79** [comment]
Small fix in change log:
(Check their in source documentation for details): replace "their" by "them"

File: source/tones.py [hunk 1]
```diff
@@ -38,18 +39,42 @@ def terminate():
...
+#: Handlers are called with the same arguments as L{beep} as keyword arguments.
+decide_beep = extensionPoints.Decider()
+
+
+def beep(
+		hz: float,
+		length: int,
+		left: int = 50,
+		right: int = 50,
+		isSPeechBeepCommand: bool = False
```
Comment (seanbudd) [inline source/tones.py:57]: there's a typo in this parameter (capital P in speech)
```suggestion
		isSpeechBeepCommand: bool = False
```

File: source/nvwave.py [hunk 2]
```diff
@@ -627,16 +642,31 @@ def outputDeviceNameToID(name: str, useDefaultIfInvalid=False) -> int:
 fileWavePlayerThread = None
 
 
-def playWaveFile(fileName, asynchronous=True):
+def playWaveFile(
+		fileName: str,
+		asynchronous: bool = True,
+		isSPeechWaveFileCommand: bool = False
```
Comment (seanbudd) [inline source/nvwave.py:648]: same typo here

Comment (leonardder) [reply source/nvwave.py:648]:
```suggestion
		isSpeechWaveFileCommand: bool = False
```

**Comment by seanbudd** [comment]
It seems that the system tests are failing for a legitimate reason on this PR

**Comment by leonardder** [comment]
I'l have a look and will add one additional extension point plus some unit tests as well.

**Comment by leonardder** [comment]
The system tests are working and all the unit tests are in place. I think this is readdy for another review.

File: source/braille.py [hunk 3]
```diff
@@ -304,6 +304,11 @@
 )
 SELECTION_SHAPE = 0xC0 #: Dots 7 and 8
 
+#: The braille shape shown on a braille display when
+#: the number of cells used by the braille handler is lower than the actual number of cells.
+#: The 0 based position of the shape is equal to the number of cells used by the braille handler.
+END_OF_BRAILLE_OUTPUT_SHAPE = 0xFF  # All dots
```
Comment (seanbudd) [inline source/braille.py:310]: For docstrings for variables, I suggest using this syntax, which is supported by Sphinx and epydoc.
The syntax used here does not give hinting in VS Code, but the epydoc syntax works in VS Code and PyCharm.

https://stackoverflow.com/questions/8820276/docstring-for-variable

```suggestion
END_OF_BRAILLE_OUTPUT_SHAPE = 0xFF  # All dots
"""
The braille shape shown on a braille display when
the number of cells used by the braille handler is lower than the actual number of cells.
The 0 based position of the shape is equal to the number of cells used by the braille handler.
"""
```

File: source/braille.py [hunk 4]
```diff
@@ -1871,11 +1876,53 @@ class BrailleHandler(baseObject.AutoPropertyObject):
 	queuedWriteLock: threading.Lock
 	ackTimerHandle: int
 
+	#: Notifies when cells are about to be written to a braille display.
+	#: This allows components and add-ons to perform an action.
+	#: For example, when a system is controlled by a braille enabled remote system,
+	#: the remote system should know what cells to show on its display.
+	#: @param cells: The list of braille cells.
+	#: @type cells: [int]
```
Comment (seanbudd) [inline source/braille.py:1884]:
```suggestion
	#: @type cells: List[int]
```

File: source/braille.py [hunk 5]
```diff
@@ -1996,7 +2076,8 @@ def setDisplayByName(  # noqa: C901
 			oldDisplay = self.display
 			if detected and bdDetect._isDebug():
 				log.debug("Possibly detected display '%s'" % newDisplay.description)
-			if newDisplay == oldDisplay.__class__:
+			sameDisplayReinit = newDisplay == oldDisplay.__class__
```
Comment (seanbudd) [inline source/braille.py:2079]: Capitalize the first I in ReInit

```suggestion
			sameDisplayReInit = newDisplay == oldDisplay.__class__
```

File: source/extensionPoints/__init__.py [hunk 6]
```diff
@@ -13,6 +13,7 @@
 from logHandler import log
 from .util import HandlerRegistrar, callWithSupportedKwargs, BoundMethodWeakref
 from typing import Set
+from typing import TypeVar, Generic
```
Comment (seanbudd) [inline source/extensionPoints/__init__.py:16]: these should be joined

```suggestion
from typing import (
	Generic,
	Set,
	TypeVar,
)
```

File: tests/unit/__init__.py [hunk 7]
```diff
@@ -91,10 +91,16 @@
 # Disable auto detection of braille displays when unit testing.
 config.conf['braille']['display'] = "noBraille"
 braille.initialize()
-# For braille unit tests, we need to construct a fake braille display as well as enable the braille handler
+
+
+# For braille unit tests, we need to enable the braille handler by providing it a  cell count
```
Comment (seanbudd) [inline tests/unit/__init__.py:96]: Fix double space

```suggestion
# For braille unit tests, we need to enable the braille handler by providing it a cell count
```

File: tests/unit/extensionPointTestHelpers.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
+# A part of NonVisual Desktop Access (NVDA)
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+# Copyright (C) 2023NV Access Limited, Babbage B.V., Leonard de Ruijter
```
Comment (seanbudd) [inline tests/unit/extensionPointTestHelpers.py:4]: fix spacing

```suggestion
# Copyright (C) 2023 NV Access Limited, Babbage B.V., Leonard de Ruijter
```

File: tests/unit/test_inputCore.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+	def tearDown(self) -> None:
+		inputCore.terminate()
+
+	def test_decide_executeGesture(self):
+		gesture = keyboardHandler.KeyboardInputGesture.fromName("NVDA+T")
+		with deciderTester(
+			self,
+			inputCore.manager.decide_executeGesture,
+			False,
```
Comment (seanbudd) [inline tests/unit/test_inputCore.py:29]: `False` is hard to understand without context. I suggest either dropping it (given the default is false), or adding the param

```suggestion
			useAssertDictContainsSubset=False,
```

File: tests/unit/test_nvwave.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	def test_decide_playWaveFile(self):
+		kwargs = {
+			"fileName": os.path.join(globalVars.appDir, "waves", "start.wav"),
+			"asynchronous": False,
+			"isSpeechWaveFileCommand": False
+		}
+		with deciderTester(
+			self,
+			nvwave.decide_playWaveFile,
+			False,
```
Comment (seanbudd) [inline tests/unit/test_nvwave.py:28]: same here re: context

File: tests/unit/test_tones.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+			"hz": 440.0,
+			"length": 500,
+			"left": 50,
+			"right": 50,
+			"isSpeechBeepCommand": False,
+		}
+		with deciderTester(
+			self,
+			tones.decide_beep,
+			False,
```
Comment (seanbudd) [inline tests/unit/test_tones.py:34]: same here re: context

File: tests/unit/test_braille.py [hunk 12]
```diff
@@ -123,3 +124,59 @@ def test_identifierWithoutModel(self):
...
+			braille.handler._displaySize,  # The currently cached display size
+			20,   # The filter handler should change the display size to 40
+		) as expectedOutput:
+			self.assertEqual(braille.handler.displaySize, expectedOutput)
+
+	def test_decide_enabled(self):
+		with deciderTester(
+			self,
+			braille.handler.decide_enabled,
+			False,
```
Comment (seanbudd) [inline tests/unit/test_braille.py:180]: same here re: context

File: source/braille.py [hunk 13]
```diff
@@ -1871,11 +1878,67 @@ class BrailleHandler(baseObject.AutoPropertyObject):
...
+
 	def __init__(self):
 		louisHelper.initialize()
 		self.display: Optional[BrailleDisplayDriver] = None
-		#: Number of cells the connected device (or if no device connected, what braille viewer has)
-		#: Zero cells disables braille. See L{_get_enabled}
+		#: Internal cache for the displaySize property.
+		#: This attribute is used to compare the displaySize output by l{filterDisplaySize}
+		#: with its previous output.
+		#: If the value differs, L{displaySizeChanged} is notified.
```
Comment (seanbudd) [inline source/braille.py:1941]: can this be turned into a docstring like the others?

**Review by seanbudd — APPROVED** [review]
Thanks @leonardder , LGTM, I can fix up the docstring when updating the changes

**Pull request merged**